### PR TITLE
WIP: New workflow syntax

### PR DESCRIPTION
**Pull request by effigies**
Work started during OHBM 2018 hackathon.

Closely following discussion in #2539. The goal is to start with @satra's syntax, and incorporate mine as a syntactic sugar. We'll see if they're equally implementable, or if there are inherent limits or affordances of each. Neither approach was fully consistent between comments, so this is somewhat an exegetical exercise.

Should merge with @djarecka's work on unifying nodes and workflows, and adding mapping and joining syntax.

### CODE REVIEW

File: nipype/pipeline/engine/workflows.py [hunk 1]
```diff
@@ -1043,3 +1044,109 @@ def _get_dot(self,
...
+class MapState(object):
+    pass
+
+class NewNode(EngineBase):
+    def __init__(self, inputs={}, map_on=None, join_by=None,
+                 *args, **kwargs):
+        self._mappers = {}
+        self._joiners = {}
+
+    def map(self, field, values=None):
```
Comment (djarecka) [inline nipype/pipeline/engine/workflows.py]: I understand that `field` is a "mapper" itself (either list, tuple or string), is that right? Not sure about `values`, do you want to be a dictionary? From the Satra example I understand that you can provide values of some of the fields from the mapper (i.e. `field` in your syntax).

File: nipype/pipeline/engine/workflows.py [hunk 2]
```diff
@@ -1043,3 +1044,109 @@ def _get_dot(self,
...
+            node = Node(runnable, name=name)
+        elif is_node(runnable):
+            node = runnable if runnable.name == name else runnable.clone(name=name)
+        else:
+            raise ValueError("Unknown workflow element: {!r}".format(runnable))
+        setattr(self, name, node)
+        self._nodes[name] = node
+        self._last_added = name
+
+    def map(self, field, node=None, values=None):
```
Comment (djarecka) [inline nipype/pipeline/engine/workflows.py]: not sure what is the meaning of `node`. it shouldn't be just `self._last_added`?

File: nipype/pipeline/engine/workflows.py [hunk 3]
```diff
@@ -1043,3 +1044,109 @@ def _get_dot(self,
...
+        self.add_nodes([node])
+
+
+class Map(Node):
+    pass
+
+
+class Join(Node):
+    pass
+
```
[the remark this answers is not in the document]
Comment (effigies) [reply nipype/pipeline/engine/workflows.py]: This is just prep for my syntax suggestion from #2539, where `Join(node, ...)` is equivalent to `node.join(...)`.

File: nipype/pipeline/engine/workflows.py (hunk 1 above)
Comment (effigies) [reply nipype/pipeline/engine/workflows.py]: I'm not actually positive of what I was doing here. Probably should have responded more promptly. :-) In any event, I clearly didn't finish this function (see L1076).

I believe what I was suggesting is that you could do:

```Python
node.map('a', [1, 2, 3, 4])

# or, where node.inputs.a takes the output of another node
node.map('a')
```

File: nipype/pipeline/engine/workflows.py (hunk 2 above)
Comment (effigies) [reply nipype/pipeline/engine/workflows.py]: Here I'm almost certain the idea is that I could do:

```Python
wf.add(node).map(<something over node>)

# or
wf.add(node)
...
wf.map(<something over node>, node=node)
```

File: nipype/pipeline/engine/workflows.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (djarecka) [reply nipype/pipeline/engine/workflows.py]: ok

File: nipype/pipeline/engine/workflows.py (hunk 2 above)
Comment (djarecka) [reply nipype/pipeline/engine/workflows.py]: oh, I see. Didn't think about this syntax.

**Comment by djarecka** [comment]
@effigies fyi: I was fixing already some exceptions etc. https://github.com/djarecka/nipype/commit/682248a7f31e51a85541b8355e4c6f58a10a53cb

File: nipype/pipeline/engine/workflows.py [hunk 4]
```diff
@@ -1043,3 +1044,109 @@ def _get_dot(self,
...
+        if node in self._mappers:
+            raise WorkflowError("Cannot assign two mappings to the same input")
+
+        self._mappers[node] = (field, values)
+
+    def join(self, field, node=None):
+        pass
+
+
+def is_function(obj):
```
Comment (kaczmarj) [inline nipype/pipeline/engine/workflows.py]: this is nitpicking, but the builtin `callable` can be used here instead of defining another function.

**Pull request closed without merging**